### ISSUE TITLE
[Search v2.2] Add support for all reportAction types in ChatListItem

### ISSUE THREAD

**Issue by luacmartins** [Expensify/App#51296]
## Problem
Right now, the search results for chats only shows `ADDCOMMENT` actions. However, we want to support all reportAction types in search.

## Solution
Refactor [ReportActionItem.tsx](https://github.com/Expensify/App/blob/3e2705055c1e8a0d20745aa3f44d72af706dcd52/src/pages/home/report/ReportActionItem.tsx) to be a pure component and receive all props it needs instead of connecting directly to Onyx, then use `ReportActionItem` in [ChatListItem](https://github.com/Expensify/App/blob/3e2705055c1e8a0d20745aa3f44d72af706dcd52/src/components/SelectionList/ChatListItem.tsx#L17) to render other reportAction types.

**Comment by stephanieelliott** [comment]
Hey @abdulrahuman5196 any update on when we can expect a PR for this?

**Comment by abdulrahuman5196** [comment]
No volunteer yet.

**Comment by luacmartins** [comment]
Still no proposals. Gonna bump the payout on this one.

**Comment by luacmartins** [comment]
@NJ-2020 thanks for the proposal. I think we need to adjust it a bit, since search data is only available within `ONYXKEYS.COLLECTION.SNAPSHOT`. We can't connect directly to any other key.

**Comment by luacmartins** [comment]
Still looking for proposals

**Comment by Pholenk** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Add support for all reportAction types in ChatListItem

### What is the root cause of that problem?
New Feature

### What changes do you think we should make in order to solve the problem?

- Since search results do not provide proper report data that is used in [**ReportActionItem.tsx**](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionItem.tsx) we need to get the detailed data based on `item.reportID` and `item.reportActionID` using the existing utils so it will look like this
```
const {reportData, reportActions, reportAction} = useMemo(() => {
        return {
            reportData: ReportUtils.getReport(item?.reportID ?? ''),
            reportActions: Object.values(ReportActionsUtils.getAllReportActions(item?.reportID ?? '')),
            reportAction: ReportActionsUtils.getReportAction(item?.reportID ?? '', item?.reportActionID ?? ''),
        };
    }, [item]);
```

- We need to replace this
https://github.com/Expensify/App/blob/e3632a9877542727230814ffc9fdd78bd087ecbd/src/components/SelectionList/ChatListItem.tsx#L80-L126

with this
```
<View style={styles.chatListWrapper}>
  <ReportActionItem
    onPress={() => onSelectRow(item)}
    report={reportData}
    reportActions={reportActions}
    parentReportAction={undefined}
    action={reportAction as ReportAction}
    displayAsGroup={false}
    isMostRecentIOUReportAction={false}
    shouldDisplayNewMarker={false}
    index={0}
    isFirstVisibleReportAction={false}
  />
</View>
```
also, we need to clean up unused lines after that.

- As you can see there is an additional wrapper to make [**ReportActionItem.tsx**](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionItem.tsx) content not overlapping the wrapper so we also need to add a new style for the wrapper called `chatListWrapper` which contains this
```
{
  flex: 1,
}
```

### What alternative solutions did you explore? (Optional)
We can refactor [**ReportActionItem.tsx**](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionItem.tsx) as mentioned earlier but it will cost a lot of effort because it already has more than 1000 lines. We may risk other features since this component is used in many pages. I don't think refactoring is the most efficient option to choose.

### Result
https://github.com/user-attachments/assets/7810c46f-34db-4886-832a-31dfb718dd2e

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by luacmartins** [comment]
>@luacmartins Could you provide more details? Is BE prepared for this feature?

Yes and no. The backend currently only returns `ADDCOMMENT` actions as part of the search result, so that part needs to be addressed in the BE. That being said, the data structure for the Search response will remain the same. We have to only use data contained in the `snapshot_` key, e.g. we can't get data from `report_`, etc keys since that's not a search result.

**Comment by wildan-m** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Add support for all reportAction types in ChatListItem

### What is the root cause of that problem?

New Feature

### What changes do you think we should make in order to solve the problem?
To minimize merge conflict and considering some `ReportActionItem` usage is similar I'd recommend to create new component that extract all onyx usage, maybe the name is `PureReportActionItem` or something else. This component will contain most of the ReportActionItem logic including memo.

Existing `ReportActionItem` will use that `PureReportActionItem` and insert onyx data.

<details><summary>ReportActionItem</summary>

```tsx

function ReportActionItem({
    action,
    report,
    ...props

}: PureReportActionItemProps) {
    const reportID = report?.reportID ?? '';
    // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
    const originalReportID = useMemo(() => ReportUtils.getOriginalReportID(reportID, action) || '-1', [reportID, action]);
    const [draftMessage] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT_ACTIONS_DRAFTS}${originalReportID}`, {
        selector: (draftMessagesForReport) => {
            const matchingDraftMessage = draftMessagesForReport?.[action.reportActionID];
            return typeof matchingDraftMessage === 'string' ? matchingDraftMessage : matchingDraftMessage?.message;
        },
    });
    const [iouReport] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT}${ReportActionsUtils.getIOUReportIDFromReportActionPreview(action) ?? -1}`);
    const [emojiReactions] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT_ACTIONS_REACTIONS}${action.reportActionID}`);
    const [userWallet] = useOnyx(ONYXKEYS.USER_WALLET);
    const [linkedTransactionRouteError] = useOnyx(
        `${ONYXKEYS.COLLECTION.TRANSACTION}${ReportActionsUtils.isMoneyRequestAction(action) ? ReportActionsUtils.getOriginalMessage(action)?.IOUTransactionID ?? -1 : -1}`,
        { selector: (transaction) => transaction?.errorFields?.route ?? null },
    );

    // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing -- This is needed to prevent the app from crashing when the app is using imported state.
    const [reportNameValuePairs] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT_NAME_VALUE_PAIRS}${report?.reportID || '-1'}`);

    const [isUserValidated] = useOnyx(ONYXKEYS.USER, { selector: (user) => !!user?.validated });
    // The app would crash due to subscribing to the entire report collection if parentReportID is an empty string. So we should have a fallback ID here.
    // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
    const [parentReport] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT}${report?.parentReportID || -1}`);


    return <PureReportActionItem
        {...props}
        action={action}
        report={report}

        draftMessage={draftMessage}
        iouReport={iouReport}
        emojiReactions={emojiReactions}
        userWallet={userWallet}
        linkedTransactionRouteError={linkedTransactionRouteError}
        reportNameValuePairs={reportNameValuePairs}
        isUserValidated={isUserValidated}
        parentReport={parentReport}


    />

```

</details>

And then this `PureReportActionItem` can also be used inside `ChatListItem` and use provided data from `snapshot_`


[Branch for this solution](https://github.com/Expensify/App/compare/main...wildan-m:wildan/fix/51296-pure-report-action-item?expand=1)
### What alternative solutions did you explore? (Optional)
N/A

**Comment by stephanieelliott** [comment]
Hey @abdulrahuman5196 couple proposals here, can you please review?

**Comment by abdulrahuman5196** [comment]
Hi will check on the proposals today

**Comment by abdulrahuman5196** [comment]
Hi, Sorry I was unable to check on this yesterday. Reaching out in slack to see if anyone can take over to avoid further delays.

**Comment by abdulrahuman5196** [comment]
@stephanieelliott Unassigning myself. Feel free to assign @allgandalf

**Comment by luacmartins** [comment]
That sounds good to me. Let's do it!

**Comment by wildan-m** [comment]
> The backend currently only returns ADDCOMMENT actions as part of the search result, so that part needs to be addressed in the BE. That being said, the data structure for the Search response will remain the same. We have to only use data contained in the snapshot_ key,

@luacmartins @allgandalf  do you think we can work in parallel? If so, can you provide me with all possible snapshot data that has not been returned from the backend yet?

**Comment by luacmartins** [comment]
I can try to work on the backend changes next week, but I think we don't need to hold for that since you already have some data to work with for the ADDCOMMENT actions

**Comment by wildan-m** [comment]
@luacmartins @allgandalf can we divide the PR into two parts?

1. Create PureReportAction and current ReportActionItem will use that with existing onyx
2. Use PureReportAction in ChatListItem

I suggest this because the `ReportActionItem` often changes, and dealing with conflicts too frequently is not very productive.

**Comment by wildan-m** [comment]
@luacmartins @allgandalf Thanks

The first part of the PRs has been created. No specific test required, just ensure that the report action item functions correctly as usual.

https://github.com/Expensify/App/pull/52948

**Comment by stephanieelliott** [comment]
PR is still under review, conversation is going on there: https://github.com/Expensify/App/pull/52948

**Comment by wildan-m** [comment]
@allgandalf yes. working on it

**Comment by wildan-m** [comment]
@luacmartins In case you've missed [my questions in this PR draft](https://github.com/Expensify/App/pull/54228#issuecomment-2547382097).

**Comment by luacmartins** [comment]
Sorry @wildan-m I was pretty busy with a HybridApp fire for the past couple of days. Replied.

**Comment by luacmartins** [comment]
@wildan-m BE PR that adds `actorAccountID` to the response is deployed

**Comment by wildan-m** [comment]
> when can the PR be ready for a review ?

@allgandalf sorry, [the new es lint rules](https://github.com/Expensify/App/pull/54228#issuecomment-2547382097) will change so many files for our case. need more time to create other two PRs

If that new rule not need to be applied, I thinks [this PR draft](https://github.com/Expensify/App/pull/54228) already completed

**Comment by wildan-m** [comment]
@allgandalf @luacmartins have you encountered this kind of error before?

https://github.com/Expensify/App/actions/runs/12478857357/job/34827066582?pr=54228

```tsx

● Test suite failed to run

    @rnmapbox/maps native code not available. Make sure you have linked the library and rebuild your app. See https://rnmapbox.github.io/docs/install?rebuild=expo#rebuild

      1 | import {useFocusEffect, useNavigation} from '@react-navigation/native';
      2 | import type {MapState} from '@rnmapbox/maps';
    > 3 | import Mapbox, {MarkerView, setAccessToken} from '@rnmapbox/maps';
        | ^
      4 | import {forwardRef, memo, useCallback, useEffect, useImperativeHandle, useMemo, useRef, useState} from 'react';
      5 | import {View} from 'react-native';
      6 | import {useOnyx} from 'react-native-onyx';


```

The error comes after putting `PureReportActionItem` inside `ChatListItem`, please take a look at `ChatListItem` change here https://github.com/Expensify/App/pull/54228/files

**Comment by wildan-m** [comment]
@allgandalf @luacmartins if we don't need to handle the lint rule related to default id, then the second PR is ready for review https://github.com/Expensify/App/pull/54228

**Comment by wildan-m** [comment]
```tsx

jest.mock('@src/components/ConfirmedRoute.tsx');

```

The above mock is to resolve jest map error mentioned here https://github.com/Expensify/App/issues/51296#issuecomment-2562168963

**Comment by stephanieelliott** [comment]
Working through PR 2/2: https://github.com/Expensify/App/pull/54228

**Comment by stephanieelliott** [comment]
PR still under active discussion/review

**Comment by luacmartins** [comment]
This PR is a beast, but we're getting through it.

**Comment by luacmartins** [comment]
@wildan-m I'd recommend reverting my revert PR so we get back to where we were and then start addressing the issues linked in the revert https://github.com/Expensify/App/pull/57360

**Comment by wildan-m** [comment]
> [@wildan-m](https://github.com/wildan-m) I'd recommend reverting my revert PR so we get back to where we were and then start addressing the issues linked in the revert [#57360](https://github.com/Expensify/App/pull/57360)

Sure will do. Working on it!

**Comment by luacmartins** [comment]
@wildan-m do you already have a draft PR?

**Comment by wildan-m** [comment]
@luacmartins yes, it's here https://github.com/Expensify/App/compare/main...wildan-m:App:wildan/fix/51296-chat-list-item-v2 I think I won't be able to finish it this week, some previously well tested cases is not working with recent merge, hope it will be completed without any regression next week

**Comment by wildan-m** [comment]
> [@wildan-m](https://github.com/wildan-m) any updates here?

I am currently working on it, but it's taking longer than expected due to the significant dependency on another component.

I need suggestion in this PR draft, please take a look https://github.com/Expensify/App/pull/57819

**Comment by wildan-m** [comment]
> Whoops! This issue is 2 days overdue. Let's get this updated quick!

This is continuously being worked on here: https://github.com/Expensify/App/pull/57819

**Comment by luacmartins** [comment]
We're making progress on the PR

**Comment by wildan-m** [comment]
@luacmartins @allgandalf There are numerous changes made during the creation of our PR, including the addition of new features. If the issues are not critical to the main flow, I believe we should avoid reverting the changes and instead opt for creating patches. This way, we can prevent the need for retesting from the beginning.

Working on creating patches....

**Comment by luacmartins** [comment]
I agree. Most of them seem minor. Let's focus on the bigger ones, like App crashing, etc.

**Comment by wildan-m** [comment]
@luacmartins Also, can we enforce—or at least make contributors aware—that any changes to ReportActionItem or its subcomponents should include verification that the ReportActionItem in the Reports section still works correctly? Otherwise, our PRs will keep getting blamed for unrelated / new features issues.

**Comment by luacmartins** [comment]
That was one of my concerns going into this feature too. I'm not really sure how to enforce that though. We can improve awareness by posting messages and maybe create a GH workflow that posts a comment to the PR whenever PureReportAction component is modified. Maybe that'd be enough to remind people to check that

**Comment by Kicu** [comment]
I think a big message should be posted on some channels, because we were quite surprised when we noticed this PR is being worked on, so my guess is not a lot of people will know.

**Comment by wildan-m** [comment]
> I think a big message should be posted on some channels

@luacmartins should I send the message?

For beta bugs, should we address them in our current issue? I believe it's important to ensure that other contributors are aware of the synchronization between live and search results in the ReportActionItem and the preferences for using a custom useOnyx wrapper.

**Comment by luacmartins** [comment]
Yea, this is handled by the overall Search v2.2 project.

**Comment by stephanieelliott** [comment]
Summarizing payment on this issue: 

 - Contributor: @wildan-m $250 via via Upwork - PAID!
 - Contributor+: @allgandalf $250 via Upwork - PAID!
 
 Upwork job is here: https://www.upwork.com/jobs/~021848820450401382223

**Comment by wildan-m** [comment]
@stephanieelliott thanks for the payment, 50% payment has been refunded
